search/search_quality/clusterize-tag-values.lisp

```diff
@@ -0,0 +1,133 @@
+#|
+exec /usr/bin/env sbcl --noinform --quit --eval "(defparameter *script-name* \"$0\")" --load "$0" --end-toplevel-options "$@"
+|#
+
+;;; This script clusterizes values from the taginfo database and
+;;; prints information about clusters.
+
+;;; Silently loads sqlite.
+(with-open-file (*standard-output* "/dev/null"
+                                   :direction :output
+                                   :if-exists :supersede)
+  (ql:quickload "sqlite"))
+
+(defun latin-char-p (char)
+  (or (and (char>= char #\a) (char<= char #\z))))
+
+(defun get-postcode-pattern (postcode)
+  "Simplifies postcode in the following way:
+   * all latin letters are replaced by 'A'
+   * all digits are replaced by 'N'
+   * hyphens and dots are replaced by a space
+   * other characters are capitalized
+
+   This format follows https://en.wikipedia.org/wiki/List_of_postal_codes.
+  "
+  (map 'string #'(lambda (c) (cond ((latin-char-p c) #\A)
+                                   ((digit-char-p c) #\N)
+                                   ((or (char= #\- c) (char= #\. c)) #\Space)
+                                   (T c)))
+       (string-upcase postcode)))
+
+
+(defun get-phone-or-flat-pattern (phone)
+  "Simplifies phone or flat numbers in the following way:
+   * all letters are replaced by 'A'
+   * all digits are replaced by 'N'
+   * other characters are capitalized
+  "
+  (map 'string #'(lambda (c) (cond ((alpha-char-p c) #\A)
+                                   ((digit-char-p c) #\N)
+                                   (T c)))
+       (string-upcase phone)))
+
+(defstruct type-settings
+  pattern-simplifier
+  field-name)
+
+(defparameter *value-type-settings*
+  `(:postcode ,(make-type-settings :pattern-simplifier #'get-postcode-pattern
+				   :field-name "addr:postcode")
+    :phone ,(make-type-settings :pattern-simplifier #'get-phone-or-flat-pattern
+				:field-name "contact:phone")
+    :flat ,(make-type-settings :pattern-simplifier #'get-phone-or-flat-pattern
+				:field-name "addr:flats")))
+
+(defstruct cluster
+  "A cluster of values with the same pattern, i.e.  all six-digits
+   series or all four-digits-two-letters series."
+  (key "") (num-samples 0) (num-occurances 0) (samples nil))
```

Comment:
num-occurences instead of num-occurances. And what's the diference between num-samples and num-occurances? You shoud use num-samples only.